Rewrite address space section.

AnastasiaStulova opened the pull request:
Discussed in internal issue 221.

This change mainly modifies existing content as follows:
- Improve structure
- Reference Embedded C and align on the terminology
- Avoid duplicating content
- Explain general rules more explicitly

Minor new content:
- Clarify address space of function return (#85)
- Clarify address space conversion in nested pointers (#30)

bashbaug commented:
Thanks for working on this - there are a lot of good changes here!

There are still some TODOs in these changes that will need to be addressed before merging.  Would you like to address these before reviewing, or would you like a first-pass review first?

I've also added this to the "OpenCL 3.0 Final specifications" project board for more visibility, but I'd like to understand if that's the right place for it or if it should move to the maintenance project board instead.  Thanks!

AnastasiaStulova commented:
> There are still some TODOs in these changes that will need to be addressed before merging. Would you like to address these before reviewing, or would you like a first-pass review first?

I would like to get some feedback from reviews first and also regarding the TODOs if possible. Thanks!

AnastasiaStulova commented:
@bashbaug I have resolved the TODOs now. There was only one that I have decided to convert into the issue https://github.com/KhronosGroup/OpenCL-Docs/issues/501 that can be addressed separately.

bashbaug commented on OpenCL_C.txt:
```diff
@@ -2335,47 +2229,394 @@ defined as the value of the <<opencl-device-queries,
+[[table-addr-spaces-summary]]
+.Address space behavior
+[%header,cols=4*]
+|====
```
I like the content of this table but I think it needs to be cleaned up a bit editorially.  I can help with this if you'd like, but a few items to consider:

1. The table cells are awkwardly sized, especially in the HTML spec.  I think this is because the HTML spec doesn't want to break the feature names.  We should consider defining asciidoc attributes with zero-width spaces to help the feature names break properly.

2. Asciidoc doesn't consider the `-` character to create bulleted lists.  I think the table would look a bit tidier with proper lists, used in all cells consistently, or removing all bulleted lists entirely.

3. I think "Addr space" is over-abbreviated.  Any reason not to spell out "Address Space"?

4. Consider using the formatting e.g. "`__global`" in the left-most column vs. "Global".

AnastasiaStulova replied:
I have addressed 2-4, but I am not sure about 1. I have found this https://stackoverflow.com/questions/40464460/how-to-do-word-break-in-ascii-doc telling that the functionality not supported. Perhaps the reason for it is that in browsers it is possible to scroll horizontally?

bashbaug commented on OpenCL_C.txt:
```diff
@@ -2335,47 +2229,394 @@ defined as the value of the <<opencl-device-queries,
-OpenCL C 2.0 or the `+__opencl_c_generic_address_space+` feature macro.
+Generic address space requires support for OpenCL C 2.0 or OpenCL C 3.0 with
+the `+__opencl_c_generic_address_space+` feature. It can be used with pointer
+types and it represents a placeholder for any of the concrete address spaces
+- `global`, `local` or `private`. It signals that a pointer points to an object
+in one of these concrete named address spaces. The exact memory segment
```
Do we need both the terms "concrete address space" and "named address spaces", or is one term sufficient?  I think we use "named address spaces" most frequently.

AnastasiaStulova replied:
Agree!

bashbaug commented on OpenCL_C.txt:
```diff
@@ -2386,14 +2627,16 @@ address space:
 
 [source,c]
 ----------
+#ifdef __opencl_c_generic_address_space
```
I'd prefer to use the comment syntax used elsewhere, especially since these examples are also valid in OpenCL C 2.0 without the `__opencl_c_generic_address_space` feature macro:

```c
// Note: these examples assume OpenCL C 2.0 or the
// __opencl_c_generic_address_space feature macro.
```

AnastasiaStulova replied:
I have changed the example to use comments instead although we could of course add a language version macro. But I think the comment conveys the message directly!  So it is better.

bashbaug commented on OpenCL_C.txt:
```diff
@@ -2482,94 +2737,102 @@ p = cp; // illegal
 // to an explicit address space pointer without a cast:
 gp = p; // compile-time error
-lp = p; // compile-time error
+lp = pi; // compile-time error
```
Was this change intended?  I believe this should remain `p` instead of `pi`.

AnastasiaStulova replied:
Yes, this is unintentional! Thanks!

bashbaug commented on OpenCL_C.txt:
```diff
@@ -2482,94 +2737,102 @@ p = cp; // illegal
+In nested pointers implicit conversions between address spaces are disallowed.
+Explicitly casting between different address spaces in nested pointers is
+allowed but the use of such pointers can lead to incorrect behavior i.e.
+accessing invalid memory sections.
```
Note: this uses the term "memory section", which isn't used anywhere else.  Should this just be "memory" instead?

AnastasiaStulova replied:
changed to "memory locations", does it make sense?

bashbaug commented on OpenCL_C.txt:
```diff
@@ -2482,94 +2737,102 @@ p = cp; // illegal
-address space and the other is a pointer in the generic address space.
+kernel void mykernel(...)
+{
+local int *local * ll;
+global int *local * gl;
+int *local * nl;
```
I think it'd be worth a few comments here to describe what's going on since this is a pretty tricky example.

AnastasiaStulova replied:
Ok, I added/changed the comment, but if you think it's not very clear I cam try to see if the example can be modified somehow...

bashbaug reviewed (commented):
Thanks - this is a really nice upgrade and it looks like it was a lot of work!

I'd like to do a minor editorial pass over these changes but I can take care of that in a separate PR after they're merged.  The comments I've left are things I wouldn't know how to fix though, so I'd appreciate addressing those before merging.  Thanks!

AnastasiaStulova commented on OpenCL_C.txt:
```diff
@@ -2335,47 +2229,394 @@ defined as the value of the <<opencl-device-queries,
---
+Private address space is a memory segment that can only be accessed by one work
+item. Variables that are not shareable among work items are allocated in private
+and it is the default address space for most of variables in particular variables
+with automatic storage duration.
```
Would it be reasonable to refer to the Inference section (i.e. addr-spaces-inference) where it is described for all variables and address spaces in details? I would prefer to avoid repeating the context in order to prevent possible inconsistency in future modifications.

AnastasiaStulova commented on OpenCL_C.txt:
```diff
@@ -2335,47 +2229,394 @@ defined as the value of the <<opencl-device-queries,
+--
+
+Address space qualifiers are not required in many cases. If they are not
+specified explicitly the default address space will be deduced depending
```
I agree, in OpenCL C it was used `to deduce` originally, but I found `to infer` was more common in language specs. Anyway I have made it consistent now.

AnastasiaStulova commented on OpenCL_C.txt:
```diff
@@ -2335,47 +2229,394 @@ defined as the value of the <<opencl-device-queries,
+{
+ private int b;
+ local int c;
+ int* p =  a ? b :  c; // p points to an object in local or private address space.
```
Definitely!

AnastasiaStulova commented on OpenCL_C.txt:
```diff
@@ -2056,73 +2056,46 @@ types.
+the region where objects are to be allocated. If the type of an
+object is qualified by an address space name, the object is allocated in the
+specified address space. Similarly pointer types can be qualified by an
+address space representing a memory segment pointed to.
```
Changed everywhere apart from:

> Private address space is a memory segment that can only be accessed by one work item.

Here I use "memory segment" as a generic concept representing physical HW rather than a language feature in OpenCL C. Do you think it is reasonable? Otherwise, I will try to change the wording.

AnastasiaStulova commented:
> Thanks - this is a really nice upgrade and it looks like it was a lot of work!
> 
> I'd like to do a minor editorial pass over these changes but I can take care of that in a separate PR after they're merged. The comments I've left are things I wouldn't know how to fix though, so I'd appreciate addressing those before merging. Thanks!

Thanks a lot for reviewing this. I have made the changes, let me know if there is anything else I can do.

alycm commented on OpenCL_C.txt:
```diff
@@ -2486,95 +2739,107 @@ lp = p; // compile-time error
+----------
 
-*Clause 6.5.6 - Additive operators, add another constraint paragraph*:
+Various clarifications and examples llustrating how changes to ISO/IEC
```
```suggestion
Various clarifications and examples illustrating how changes to ISO/IEC
```

alycm commented on OpenCL_C.txt:
```diff
@@ -2486,95 +2739,107 @@ lp = p; // compile-time error
-Conversion of a null pointer to another pointer type yields a null pointer
-of that type.
-Any two null pointers whose referenced address spaces overlap shall compare
-equal.
```
NOT EDITORIAL.

AnastasiaStulova replied:
The second sentence is part of the Embedded C spec, section 5.3:
>  Any two null pointers whose referenced address spaces overlap shall compare equal.

I don't believe OpenCL changes this.

Btw, the first sentence doesn't seem to follow the Embedded C strickly but I am trying to understand how this could be implemented since we don't have `NULL` pointer constant for different address spaces? I was trying to think of a code example of this behavior:

__private int* p1 = NULL; // no conversion, NULL is in __private address space here
__local int* p2 = NULL; // no conversion, NULL is in __local address space here
int* gen = p1; // the parser doesn't know what values p1 has, so it can only perform regular address space conversion.

alycm replied:
> The second sentence is part of the Embedded C spec, section 5.3:

Ah, of course.

For the first sentence, going back to C99, it says:

> Conversion  of  a  null  pointer  to  another  pointer  type  yields  a  null  pointer  of  that  type.  Any two null pointers shall compare equal.

So the only change here is w.r.t. to comparing NULL, and your quote from Embedded C already covers that.

Thanks for the clarification, no further action required, I'll resolve this.

alycm commented on OpenCL_C.txt:
```diff
@@ -2335,47 +2229,394 @@ defined as the value of the <<opencl-device-queries,
+qualifier.
+
+Variables declared in the outermost compound statement inside the body of the
+kernel function can be qualified by local or constant address spaces.
```
NOT EDITORIAL.

AnastasiaStulova replied:
Section 6.5.3 of OpenCL C 2.0

https://www.khronos.org/registry/OpenCL/specs/2.2/html/OpenCL_C.html#constant-or-constant

>Variables in the outermost scope of kernel functions can be declared in the __constant address space.

alycm replied:
Ah, thanks!  Resolving this.

alycm reviewed (approved):
LGTM.  AFAICT this doesn't change the meaning of an OpenCL C program.

The vast majority of my comments are editorial and can be fixed post-merge (@bashbaug said something similar).

I've added a "NOT EDITORIAL" comment to two of them.

bashbaug commented:
The biggest issue with this PR is that it needs a rebase before merging.  I tried to do this last night but it's not trivial...

AnastasiaStulova commented on OpenCL_C.txt:
```diff
@@ -2335,47 +2229,394 @@ defined as the value of the <<opencl-device-queries,
+on the declaration scope and the object type.
+
+There is no syntax to provide address space in the source for some situation,
+therefore only default address space is applicable.
```
String literal or temporaries. I mention string literals below but temporaries are a bit harder because C doesn't define that concept well. I am thinking perhaps we should commit this sentence since it is not very clear and not easy to explain either?

AnastasiaStulova commented on OpenCL_C.txt:
```diff
@@ -2056,73 +2056,46 @@ types.
-to the generic address space.
-Casting a pointer to address space A to a pointer to address space B is
-illegal if A and B are named address spaces and A is not the same as B.
+OpenCL C relaxes some restriction from <<embedded-c-spec, section 5.1.2 and
```
I think in OpenCL we relax the restriction for address spaces in non-local variables i.e. at program scope?

alycm replied:
This sentence doesn't say that though.

I think I'm not clear on what this sentence intends to say.  Some of what it says is repeating embedded C, the part after the "but" is overriding part of it I think.   That's a lot to try and pack into one sentence.

AnastasiaStulova replied:
Ok, I see. Maybe we should change this to something like:

`
Most of the restrictions from <<embedded-c-spec, section 5.1.2 and section 5.3 apply in OpenCL. C i.e. address spaces can not be used with a return type, a function parameter, or a function type; multiple address space qualifiers are not allowed. However, in OpenCL C it is allowed to qualify local variables with an address space qualifier.
`

alycm replied:
Yes, that sounds clearer.

AnastasiaStulova replied:
Ok, this has been updated. Thanks!

AnastasiaStulova commented on OpenCL_C.txt:
```diff
@@ -2335,47 +2229,394 @@ defined as the value of the <<opencl-device-queries,
+This section describes use of address space qualifiers with respect to
+declaration scopes or variable types.
+
+Local variables inside functions can be qualifier by private address space
```
Although we can also declare variables inside functions that are not local i.e. `static` or `extern` variables?

AnastasiaStulova commented on OpenCL_C.txt:
```diff
@@ -2056,73 +2056,46 @@ types.
 {
     // declares an array of 4 floats in the private address space
-    float x[4];
+    private float x[4];
```
Fixed! Thanks!

AnastasiaStulova commented on OpenCL_C.txt:
```diff
@@ -2056,73 +2056,46 @@ types.
 // declares a pointer p in the global address space that
 // points to an object in the global address space
-global int *p;
+__global int *global p;
```
Fixed! Thanks!

AnastasiaStulova commented:
> The biggest issue with this PR is that it needs a rebase before merging. I tried to do this last night but it's not trivial...

Yes, it was a nightmare to rebase. I had to reapply the changes on top of the current matser. That ended up easier. :)

AnastasiaStulova commented:
I think I have addressed most of the comments now. If there are no further requests I would like to merge this soon to avoid nasty rebases. Happy to improve the wording if needed in subsequent PRs.
